File: cpp/algorithms/string/ac_automaton.cpp
// Aho–Corasick 自动机（多模式匹配）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ACAutomaton {
    struct Node {
        array<int, 26> next{};
        int fail = 0;
        vector<pair<int, int>> out;  // (pattern_id, length)
        Node() { next.fill(-1); }
    };
    vector<Node> nodes{Node()};

    int idx(char c) { return c - 'a'; }

    void add(const string& pat, int pid) {
        if (pat.empty()) throw runtime_error("empty pattern not supported");
        int u = 0;
        for (char c : pat) {
            int id = idx(c);
            if (nodes[u].next[id] == -1) {
                nodes[u].next[id] = (int)nodes.size();
                nodes.emplace_back();
            }
            u = nodes[u].next[id];
        }
        nodes[u].out.emplace_back(pid, (int)pat.size());
    }

    void build() {
        queue<int> q;
        for (int c = 0; c < 26; ++c) {
            int v = nodes[0].next[c];
            if (v != -1) {
                nodes[v].fail = 0;
                q.push(v);
            }
        }
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            for (int c = 0; c < 26; ++c) {
                int v = nodes[u].next[c];
                if (v == -1) continue;
                q.push(v);
                int f = nodes[u].fail;
                while (f && nodes[f].next[c] == -1) f = nodes[f].fail;
                nodes[v].fail = nodes[f].next[c] == -1 ? 0 : nodes[f].next[c];
                auto& ov = nodes[v].out;
                auto& of = nodes[nodes[v].fail].out;
                ov.insert(ov.end(), of.begin(), of.end());
            }
        }
    }

    vector<array<int, 3>> find_all(const string& text) {
        vector<array<int, 3>> res;
        int state = 0;
        for (int i = 0; i < (int)text.size(); ++i) {
            int c = idx(text[i]);
            while (state && nodes[state].next[c] == -1) state = nodes[state].fail;
            if (nodes[state].next[c] != -1) state = nodes[state].next[c];
            for (auto [pid, ln] : nodes[state].out) {
                res.push_back({i - ln + 1, pid, ln});
            }
        }
        return res;
    }
};

int main() {
    vector<string> pats{"he", "she", "his", "hers"};
    ACAutomaton ac;
    for (int i = 0; i < (int)pats.size(); ++i) ac.add(pats[i], i);
    ac.build();
    auto hits = ac.find_all("ushers");
    vector<pair<int, int>> got, want{{1, 1}, {2, 0}, {2, 3}};
    for (auto& h : hits) got.push_back({h[0], h[1]});
    sort(got.begin(), got.end());
    sort(want.begin(), want.end());
    assert(got == want);
    try {
        ac.add("", 0);
        assert(false);
    } catch (const runtime_error&) {
    }
    cout << "ac_automaton OK" << endl;
    return 0;
}

File: python/algorithms/string/ac_automaton.py
"""Aho–Corasick 自动机：多模式串线性扫描匹配。"""

from __future__ import annotations

from collections import deque
from typing import Dict, List, Tuple


class ACAutomaton:
    def __init__(self) -> None:
        self._next: List[Dict[str, int]] = [{}]
        self._fail: List[int] = [0]
        self._out: List[List[Tuple[int, int]]] = [[]]  # (pattern_id, length)

    def add(self, pat: str, pid: int) -> None:
        if not pat:
            raise ValueError("empty pattern not supported")
        node = 0
        for ch in pat:
            if ch not in self._next[node]:
                self._next[node][ch] = len(self._next)
                self._next.append({})
                self._fail.append(0)
                self._out.append([])
            node = self._next[node][ch]
        self._out[node].append((pid, len(pat)))

    def build(self) -> None:
        q: deque[int] = deque()
        for ch, nxt in self._next[0].items():
            self._fail[nxt] = 0
            q.append(nxt)
        while q:
            u = q.popleft()
            for ch, v in self._next[u].items():
                q.append(v)
                f = self._fail[u]
                while f and ch not in self._next[f]:
                    f = self._fail[f]
                self._fail[v] = self._next[f][ch] if ch in self._next[f] else 0
                self._out[v].extend(self._out[self._fail[v]])

    def find_all(self, text: str) -> List[Tuple[int, int, int]]:
        """返回 (起始下标, pattern_id, 长度) 列表。"""
        res: List[Tuple[int, int, int]] = []
        state = 0
        for i, ch in enumerate(text):
            while state and ch not in self._next[state]:
                state = self._fail[state]
            if ch in self._next[state]:
                state = self._next[state][ch]
            for pid, ln in self._out[state]:
                res.append((i - ln + 1, pid, ln))
        return res


if __name__ == "__main__":
    pats = ["he", "she", "his", "hers"]
    ac = ACAutomaton()
    for i, p in enumerate(pats):
        ac.add(p, i)
    ac.build()
    hits = ac.find_all("ushers")
    got = sorted((pos, pid) for pos, pid, _ in hits)
    want = sorted([(1, 1), (2, 0), (2, 3)])
    assert got == want
    try:
        ac.add("", 0)
        raise AssertionError("expected ValueError")
    except ValueError:
        pass
    print("ac_automaton OK")
